flask_pymongo/__init__.py

```diff
@@ -237,5 +238,19 @@ def save_upload(filename):
             db_obj = self.db
         assert db_obj is not None, "Please initialize the app before calling save_file!"
         storage = GridFS(db_obj, base)
-        id = storage.put(fileobj, filename=filename, content_type=content_type, **kwargs)
+
+        # GridFS does not manage its own hash, so we manage our own using its
+        # metadata storage, to be used for the etag.
+        sha1_sum = self._compute_sha(fileobj)
+        metadata = dict(sha1_sum=sha1_sum)
+        id = storage.put(
+            fileobj, filename=filename, content_type=content_type, metadata=metadata, **kwargs
+        )
         return id
+
+    def _compute_sha(self, fileobj: Any) -> str:
+        """Compute the sha sum of a file object."""
+        pos = fileobj.tell()
+        raw_data = fileobj.read()
+        fileobj.seek(pos)
+        return hashlib.sha1(raw_data).hexdigest()
```

Comment:
Instead of reading the whole file, it's better to iterate the file data in chunks and calculate the hash:
```python
hash = hashlib.sha1()
while True:
    chunk = fileobj.readchunk()
    if not chunk:
        break
    hash.update(chunk)
sha1_sum = hash.hexdigest()
```

That way we don't have to assemble the entire file in memory at once. However, I suggest we *don't* implement this optimization now and defer it as future work because there are too many other questions in flight that would change this code.